### PR TITLE
Enable etcd tls

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
Enable tls for client -> etcd communication.
Client certificate authentication will be handled in a follow-up PR

**Release note**:
```release-note
Clients will now talk to etcd over TLS
```

### CODE REVIEW

File: api/pkg/controller/cluster/resources.go [hunk 1]
```diff
@@ -69,6 +69,11 @@ func (cc *Controller) ensureResourcesAreDeployed(cluster *kubermaticv1.Cluster)
 		return err
 	}
 
+	// check that all secrets are available // New way of handling secrets
+	if err := cc.newEnsureSecrets(cluster); err != nil {
```
Comment (alvaroaleman) [inline api/pkg/controller/cluster/resources.go]: What about `EnsureSecretsV2` so we avoid having at some point `newNewNewEnsureSecrets`?

File: api/pkg/resources/etcd/tls-serving-certificate.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	altNames := sets.NewString()
+	for i := 0; i < 3; i++ {
+		// Member name
+		podName := fmt.Sprintf("etcd-%d", i)
+		altNames.Insert(podName)
+
+		// Pod DNS name
+		absolutePodDNSName := fmt.Sprintf("etcd-%d.%s.%s.svc.cluster.local", i, resources.EtcdServiceName, data.Cluster.Status.NamespaceName)
```
Comment (alvaroaleman) [inline api/pkg/resources/etcd/tls-serving-certificate.go:39]: Maybe sign for `127.0.0.1` as well?

File: api/pkg/resources/etcd/tls-serving-certificate.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+	newKP, err := triple.NewServerKeyPair(ca, "etcd", resources.EtcdClientServiceName, data.Cluster.Status.NamespaceName, "cluster.local", []string{clientIP}, altNames.List())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create apiserver key pair: %v", err)
+	}
+
+	if b, exists := se.Data[resources.EtcdTLSCertSecretKey]; exists {
+		certs, err := certutil.ParseCertsPEM(b)
+		if err != nil {
+			return nil, fmt.Errorf("faield to certificate (key=%s) from existing secret %s: %v", resources.EtcdTLSCertSecretKey, resources.EtcdTLSCertificateSecretName, err)
```
Comment (thz) [inline api/pkg/resources/etcd/tls-serving-certificate.go]: s/faield/failed/

Comment (thz) [reply api/pkg/resources/etcd/tls-serving-certificate.go]: also "create" missing

Comment (mrIncompetent) [reply api/pkg/resources/etcd/tls-serving-certificate.go]: fixed with https://github.com/kubermatic/kubermatic/pull/1495/commits/1c49032fc2c04109b77930449b05b787d3f1be5b

File: api/pkg/resources/etcd/tls-serving-certificate.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/resources/etcd/tls-serving-certificate.go:39]: fixed with https://github.com/kubermatic/kubermatic/pull/1495/commits/1c49032fc2c04109b77930449b05b787d3f1be5b

File: api/pkg/controller/cluster/resources.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/controller/cluster/resources.go]: fixed with https://github.com/kubermatic/kubermatic/pull/1495/commits/1c49032fc2c04109b77930449b05b787d3f1be5b

**Comment by mrIncompetent** [comment]
Need to fix the certificate validation first.
Current approach is far to cpu heavy

File: api/pkg/resources/resources.go [hunk 4]
```diff
@@ -322,7 +334,64 @@ func (d *TemplateData) ImageRegistry(defaultRegistry string) string {
...
+	}, nil
+}
+
 // GetLabels returns default labels every resource should have
 func GetLabels(app string) map[string]string {
 	return map[string]string{AppLabelKey: app}
 }
+
+// ServerCertificateNeedsUpdate validates if the given data is present in the given certificate
+func ServerCertificateNeedsUpdate(cert *x509.Certificate, commonName, svcName, svcNamespace, dnsDomain string, ips, hostnames []string) bool {
```
Comment (alvaroaleman) [inline api/pkg/resources/resources.go]: I usually find it easier to read if the logic is not negated, e.G. here use `IsCertificateValidForAllOf`

Also, wouldn't it be easier to adjust the signature to just take `cert *x509.Certificate, ips, hostNames []string`? I find the current signature rather confusing

**Review by alvaroaleman — CHANGES_REQUESTED** [review]

File: api/pkg/resources/resources.go (hunk 4 above)
Comment (mrIncompetent) [reply api/pkg/resources/resources.go]: Its the same signature (except `cert *x509.Certificate`) the kubernetes-client offers to create certificates.
I would rather stick to it, unless you have a strong opinion in changing it

File: api/pkg/resources/etcd/tls-serving-certificate.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		return nil, fmt.Errorf("failed to get ClientIP of etcd client service '%s': %v", resources.EtcdClientServiceName, err)
+	}
+
+	if b, exists := se.Data[resources.EtcdTLSCertSecretKey]; exists {
+		certs, err := certutil.ParseCertsPEM(b)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse certificate (key=%s) from existing secret %s: %v", resources.EtcdTLSCertSecretKey, resources.EtcdTLSCertificateSecretName, err)
+		}
+
+		if !resources.ServerCertificateNeedsUpdate(certs[0], "etcd", resources.EtcdClientServiceName, data.Cluster.Status.NamespaceName, "cluster.local", []string{clientIP}, altNames.List()) {
```
Comment (alvaroaleman) [inline api/pkg/resources/etcd/tls-serving-certificate.go]: We should probably also check if its going to expire soon

Comment (mrIncompetent) [reply api/pkg/resources/etcd/tls-serving-certificate.go]: Something i'm already doing in a follow up

**Review by alvaroaleman — DISMISSED** [review]

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**